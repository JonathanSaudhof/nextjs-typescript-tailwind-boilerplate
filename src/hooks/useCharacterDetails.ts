import useSWR from "swr";
import fetcher from "@/lib/fetcher";

import { extractIdFromUrl, prefetcher } from "@/lib/helper";
import { CustomError } from "@/lib/customError";

export interface CharacterRawDetails {
  birth_year: string;
  created: string;
  edited: string;
  eye_color: string;
  films: string[];
  gender: string;
  hair_color: string;
  height: string;
  homeworld: string;
  mass: string;
  name: string;
  skin_color: string;
  species: string[];
  starships: string[];
  url: string;
  vehicles: string[];
}

interface CharacterDetails {
  name: string;
  hairColor: string;
  skinColor: string;
  birthYear: string;
  gender: string;
  films: number[];
  species: string[];
}

export function useCharacterDetails(id: number | null): {
  characterDetails: CharacterDetails | null;
  isLoading: boolean;
  isError: CustomError;
} {
  const { data, error } = useSWR<CharacterRawDetails>(
    id ? `https://swapi.dev/api/people/${id}` : null,
    fetcher,
  );

  if (data) {
    const {
      name,
      hair_color: hairColor,
      skin_color: skinColor,
      birth_year: birthYear,
      gender,
      films: rawFilms, // ids need to be extracted
      species,
    } = data;

    const films = rawFilms.map((film) => extractIdFromUrl(film));

    return {
      characterDetails: {
        name,
        hairColor,
        skinColor,
        birthYear,
        gender,
        films,
        species,
      },
      isLoading: false,
      isError: error,
    };
  }

  return {
    characterDetails: null,
    isLoading: true,
    isError: error,
  };
}

export async function prefetchCharacterDetails(id: number) {
  const url = `https://swapi.dev/api/people/?${id}`;
  prefetcher(url);
}
