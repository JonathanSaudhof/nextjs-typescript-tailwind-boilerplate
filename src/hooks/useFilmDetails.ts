import useSWR from "swr";
import fetcher from "@src/lib/fetcher";
import { prefetcher } from "@lib/helper";

interface FilmRawDetails {
  id: number;
  episodeId: string;
}

export default function useFilmDetails(id: number): {
  filmDetails: FilmRawDetails;
  isLoading: boolean;
  isError: Error;
} {
  const { data, error } = useSWR(`https://swapi.dev/api/films/${id}`, fetcher);

  if (data) {
    const { count, results: characterRawList } = data;
    numberOfPages = Math.round(+count / ITEMS_PER_PAGE);
    characterList = extractCharacterListItemInformation(characterRawList);

    return {
      characterList,
      currentPage: page,
      pages: numberOfPages,
      isLoading: false,
      isError: error,
    };
  }

  return {
    characterList: [],
    currentPage: page,
    pages: numberOfPages,
    isLoading: true,
    isError: error,
  };
}

export function prefetchFilms(id: number) {
  const url = `https://swapi.dev/api/films/${id}/`;
  prefetcher(url);
}
