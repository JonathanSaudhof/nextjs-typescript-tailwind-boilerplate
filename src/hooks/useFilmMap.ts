import useSWR from "swr";
import fetcher from "@src/lib/fetcher";
import { extractIdFromUrl } from "@lib/helper";

interface FilmRawDetails {
  episode_id: string;
  title: string;
  opening_crawl: string;
  director: string;
  producer: string;
  characters: string[];
  species: string[];
  vehicles: string[];
  starships: string[];
  planets: string[];
  release_date: string;
  created: string;
  edited: string;
  url: string;
}

interface FilmRawList {
  results: FilmRawDetails[];
  count: number;
  next: string;
  previous: string;
}

interface FilmDetails {
  title: string;
  episodeId: string;
  openingCrawl: string;
  director: string;
  producer: string;
  characters: number[];
  releaseDate: string;
}

interface FilmMap {
  [id: number]: FilmDetails;
}

export default function useFilmMap(): {
  filmMap: FilmMap | null;
  isLoading: boolean;
  isError: Error;
} {
  const { data, error } = useSWR<FilmRawList>(
    `https://swapi.dev/api/films`,
    fetcher,
  );

  if (data) {
    const { results: filmRawList } = data;

    let filmMap: FilmMap = {};

    for (const film of filmRawList) {
      const id = extractIdFromUrl(film.url);
      const characters = film.characters.map((character) =>
        extractIdFromUrl(character),
      );

      filmMap[id] = {
        title: film.title,
        episodeId: film.episode_id,
        openingCrawl: film.opening_crawl,
        director: film.director,
        producer: film.producer,
        characters,
        releaseDate: film.release_date,
      };
    }

    return {
      filmMap,
      isLoading: false,
      isError: error,
    };
  }

  return {
    filmMap: null,
    isLoading: true,
    isError: error,
  };
}
