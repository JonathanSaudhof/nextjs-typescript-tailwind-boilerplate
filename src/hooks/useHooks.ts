export function useHooks() {
  // use Hooks in here
}
