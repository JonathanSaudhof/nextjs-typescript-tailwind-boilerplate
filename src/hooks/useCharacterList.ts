import useSWR from "swr";
import fetcher from "@/lib/fetcher";
import { extractIdFromUrl, prefetcher } from "@/lib/helper";
import { CharacterRawDetails } from "./useCharacterDetails";

const ITEMS_PER_PAGE = 10;

export type CharcerRawList = {
  count: number;
  next: string;
  previous: string | null;
  results: CharacterRawDetails[];
};

interface CharcterDetails {
  id: number;
  name: string;
  gender: string;
  species: string[];
  birthYear: string;
  eyeColor: string;
  films: number[];
  url: string;
}

export function useCharacterList(page: number): {
  characterList: CharcterDetails[] | [];
  pages: number | null;
  isLoading: boolean;
  isError: Error;
} {
  const { data, error } = useSWR<CharcerRawList | undefined>(
    page ? `https://swapi.dev/api/people/?page=${page}` : null,
    fetcher,
  );

  let numberOfPages: number | null = null;
  let characterList: any[] = [];

  if (data) {
    const { count, results: characterRawList } = data;
    numberOfPages = Math.round(+count / ITEMS_PER_PAGE);
    characterList = extractCharacterListItemInformation(characterRawList);

    return {
      characterList,
      pages: numberOfPages,
      isLoading: false,
      isError: error,
    };
  }

  return {
    characterList: [],
    pages: numberOfPages,
    isLoading: true,
    isError: error,
  };
}

/**
 * extractCharacterListInformation
 * @description Get a subset of data from the raw characters list retruned from the api
 * to be shown in the list view
 */
function extractCharacterListItemInformation(
  characterRawList: CharacterRawDetails[],
): CharcterDetails[] {
  return characterRawList.map((character) => {
    const id = extractIdFromUrl(character.url);

    const films = character.films.map((film) => {
      const id = extractIdFromUrl(film);

      return id;
    });

    return {
      id,
      name: character.name,
      gender: character.gender,
      species: character.species,
      birthYear: character.birth_year,
      eyeColor: character.eye_color,
      films,
      url: character.url,
    };
  });
}

export async function prefetchCharacterList(page: number) {
  const url = `https://swapi.dev/api/people/?page=${page}`;
  prefetcher(url);
}
