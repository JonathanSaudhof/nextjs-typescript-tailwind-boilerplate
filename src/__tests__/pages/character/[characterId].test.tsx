import CharacterDetails from "@/pages/character/[characterId]";
import { render, cleanup } from "@testing-library/react";
import { works, fails } from "../../__mocks__/mockCharacterDetails";
import filmMap from "../../__mocks__/mockFilmMap";

const useRouter = jest.spyOn(require("next/router"), "useRouter");

const useCharacterDetails = jest.spyOn(
  require("@/hooks/useCharacterDetails"),
  "useCharacterDetails",
);

jest.mock("@/hooks/useFilmMap", () => ({
  useFilmMap: () => filmMap,
}));

describe("Character Details Page", () => {
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  // Happy Path
  it("should render detail of character with id 1", async () => {
    const mockRouter = {
      query: {
        characterId: "1",
      },
    };

    useRouter.mockReturnValue(mockRouter);
    useCharacterDetails.mockImplementation(() => works);

    const { asFragment } = render(<CharacterDetails />);
    expect(useCharacterDetails).toHaveBeenCalledWith(1);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render detail of character with id 2", async () => {
    const mockRouter = {
      query: {
        characterId: "2",
      },
    };

    useRouter.mockReturnValue(mockRouter);
    useCharacterDetails.mockImplementationOnce(() => works);

    const { asFragment } = render(<CharacterDetails />);
    expect(useCharacterDetails).toHaveBeenCalledWith(2);
    expect(asFragment()).toMatchSnapshot();
  });

  // Wrong id
  it("should forward to 404 on wrong id", async () => {
    const mockRouter = {
      query: {
        characterId: "asdf",
      },
      push: jest.fn(),
    };

    useRouter.mockReturnValue(mockRouter);

    useCharacterDetails.mockImplementationOnce(() => fails);
    render(<CharacterDetails />);
    expect(useCharacterDetails).toHaveBeenCalledWith(null);
    expect(mockRouter.push).toHaveBeenCalledWith("/404");
  });
});
