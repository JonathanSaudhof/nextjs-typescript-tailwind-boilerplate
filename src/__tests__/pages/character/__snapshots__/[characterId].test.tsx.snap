// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character Details Page should render detail of character with id 1 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto py-20"
  >
    <h1>
      Luke Skywalker
    </h1>
    <p>
      male
    </p>
    <p>
      19BBY
    </p>
    <p>
      blond
    </p>
    <p
      class=""
    >
      Films
    </p>
    <ul
      class="gap-6"
    >
      <li
        class="py-6 flex gap-3"
      >
        <div>
          4
        </div>
        <div>
          A New Hope
        </div>
        <div>
          George Lucas
        </div>
        <div>
          It is a period of civil war.
Rebel spaceships, striking
from a hidden base, have won
their first victory against
the evil Galactic Empire.

During the battle, Rebel
spies managed to steal secret
plans to the Empire's
ultimate weapon, the DEATH
STAR, an armored space
station with enough power
to destroy an entire planet.

Pursued by the Empire's
sinister agents, Princess
Leia races home aboard her
starship, custodian of the
stolen plans that can save her
people and restore
freedom to the galaxy....
        </div>
      </li>
      <li
        class="py-6 flex gap-3"
      >
        <div>
          5
        </div>
        <div>
          The Empire Strikes Back
        </div>
        <div>
          Irvin Kershner
        </div>
        <div>
          It is a dark time for the
Rebellion. Although the Death
Star has been destroyed,
Imperial troops have driven the
Rebel forces from their hidden
base and pursued them across
the galaxy.

Evading the dreaded Imperial
Starfleet, a group of freedom
fighters led by Luke Skywalker
has established a new secret
base on the remote ice world
of Hoth.

The evil lord Darth Vader,
obsessed with finding young
Skywalker, has dispatched
thousands of remote probes into
the far reaches of space....
        </div>
      </li>
      <li
        class="py-6 flex gap-3"
      >
        <div>
          6
        </div>
        <div>
          Return of the Jedi
        </div>
        <div>
          Richard Marquand
        </div>
        <div>
          Luke Skywalker has returned to
his home planet of Tatooine in
an attempt to rescue his
friend Han Solo from the
clutches of the vile gangster
Jabba the Hutt.

Little does Luke know that the
GALACTIC EMPIRE has secretly
begun construction on a new
armored space station even
more powerful than the first
dreaded Death Star.

When completed, this ultimate
weapon will spell certain doom
for the small band of rebels
struggling to restore freedom
to the galaxy...
        </div>
      </li>
      <li
        class="py-6 flex gap-3"
      >
        <div>
          3
        </div>
        <div>
          Revenge of the Sith
        </div>
        <div>
          George Lucas
        </div>
        <div>
          War! The Republic is crumbling
under attacks by the ruthless
Sith Lord, Count Dooku.
There are heroes on both sides.
Evil is everywhere.

In a stunning move, the
fiendish droid leader, General
Grievous, has swept into the
Republic capital and kidnapped
Chancellor Palpatine, leader of
the Galactic Senate.

As the Separatist Droid Army
attempts to flee the besieged
capital with their valuable
hostage, two Jedi Knights lead a
desperate mission to rescue the
captive Chancellor....
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Character Details Page should render detail of character with id 2 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto py-20"
  >
    <h1>
      Luke Skywalker
    </h1>
    <p>
      male
    </p>
    <p>
      19BBY
    </p>
    <p>
      blond
    </p>
    <p
      class=""
    >
      Films
    </p>
    <ul
      class="gap-6"
    >
      <li
        class="py-6 flex gap-3"
      >
        <div>
          4
        </div>
        <div>
          A New Hope
        </div>
        <div>
          George Lucas
        </div>
        <div>
          It is a period of civil war.
Rebel spaceships, striking
from a hidden base, have won
their first victory against
the evil Galactic Empire.

During the battle, Rebel
spies managed to steal secret
plans to the Empire's
ultimate weapon, the DEATH
STAR, an armored space
station with enough power
to destroy an entire planet.

Pursued by the Empire's
sinister agents, Princess
Leia races home aboard her
starship, custodian of the
stolen plans that can save her
people and restore
freedom to the galaxy....
        </div>
      </li>
      <li
        class="py-6 flex gap-3"
      >
        <div>
          5
        </div>
        <div>
          The Empire Strikes Back
        </div>
        <div>
          Irvin Kershner
        </div>
        <div>
          It is a dark time for the
Rebellion. Although the Death
Star has been destroyed,
Imperial troops have driven the
Rebel forces from their hidden
base and pursued them across
the galaxy.

Evading the dreaded Imperial
Starfleet, a group of freedom
fighters led by Luke Skywalker
has established a new secret
base on the remote ice world
of Hoth.

The evil lord Darth Vader,
obsessed with finding young
Skywalker, has dispatched
thousands of remote probes into
the far reaches of space....
        </div>
      </li>
      <li
        class="py-6 flex gap-3"
      >
        <div>
          6
        </div>
        <div>
          Return of the Jedi
        </div>
        <div>
          Richard Marquand
        </div>
        <div>
          Luke Skywalker has returned to
his home planet of Tatooine in
an attempt to rescue his
friend Han Solo from the
clutches of the vile gangster
Jabba the Hutt.

Little does Luke know that the
GALACTIC EMPIRE has secretly
begun construction on a new
armored space station even
more powerful than the first
dreaded Death Star.

When completed, this ultimate
weapon will spell certain doom
for the small band of rebels
struggling to restore freedom
to the galaxy...
        </div>
      </li>
      <li
        class="py-6 flex gap-3"
      >
        <div>
          3
        </div>
        <div>
          Revenge of the Sith
        </div>
        <div>
          George Lucas
        </div>
        <div>
          War! The Republic is crumbling
under attacks by the ruthless
Sith Lord, Count Dooku.
There are heroes on both sides.
Evil is everywhere.

In a stunning move, the
fiendish droid leader, General
Grievous, has swept into the
Republic capital and kidnapped
Chancellor Palpatine, leader of
the Galactic Senate.

As the Separatist Droid Army
attempts to flee the besieged
capital with their valuable
hostage, two Jedi Knights lead a
desperate mission to rescue the
captive Chancellor....
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
